File: WellnessWingman/Services/Navigation/HistoricalNavigationContext.cs
using System;
using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;

namespace HealthHelper.Services.Navigation;

/// <summary>
/// Tracks the currently displayed historical view and maintains the breadcrumb stack
/// used to restore the previous view when navigating backwards.
/// </summary>
public partial class HistoricalNavigationContext : ObservableObject
{
    private readonly ObservableCollection<NavigationBreadcrumb> _breadcrumbStack = new();
    private HistoricalViewLevel currentLevel;
    private DateTime currentDate;

    public HistoricalNavigationContext()
    {
        currentLevel = HistoricalViewLevel.Today;
        currentDate = DateTime.Today;
        Breadcrumbs = new ReadOnlyObservableCollection<NavigationBreadcrumb>(_breadcrumbStack);
    }

    /// <summary>
    /// Gets a read-only view of the breadcrumb stack with the oldest item at index 0.
    /// </summary>
    public ReadOnlyObservableCollection<NavigationBreadcrumb> Breadcrumbs { get; }

    public HistoricalViewLevel CurrentLevel
    {
        get => currentLevel;
        private set => SetProperty(ref currentLevel, value);
    }

    public DateTime CurrentDate
    {
        get => currentDate;
        private set => SetProperty(ref currentDate, value);
    }

    public bool HasBreadcrumbs => _breadcrumbStack.Count > 0;

    /// <summary>
    /// Creates a breadcrumb for the current state so the caller can push it before changing levels.
    /// </summary>
    public NavigationBreadcrumb CreateCurrentBreadcrumb(string? label = null)
    {
        return new NavigationBreadcrumb(CurrentLevel, CurrentDate, label);
    }

    /// <summary>
    /// Returns the most recent breadcrumb without modifying the stack.
    /// </summary>
    public NavigationBreadcrumb? PeekBreadcrumb()
    {
        if (_breadcrumbStack.Count == 0)
        {
            return null;
        }

        return _breadcrumbStack[^1];
    }

    /// <summary>
    /// Pushes a breadcrumb representing the view the user is drilling down from.
    /// </summary>
    /// <remarks>The most recent breadcrumb is always the top of the stack.</remarks>
    public void PushBreadcrumb(NavigationBreadcrumb breadcrumb)
    {
        _breadcrumbStack.Add(breadcrumb);
        OnPropertyChanged(nameof(HasBreadcrumbs));
    }

    /// <summary>
    /// Pops the most recent breadcrumb when the user navigates back up the hierarchy.
    /// </summary>
    /// <returns>The previously active breadcrumb or <c>null</c> if the stack is empty.</returns>
    public NavigationBreadcrumb? PopBreadcrumb()
    {
        if (_breadcrumbStack.Count == 0)
        {
            return null;
        }

        var index = _breadcrumbStack.Count - 1;
        var breadcrumb = _breadcrumbStack[index];
        _breadcrumbStack.RemoveAt(index);
        OnPropertyChanged(nameof(HasBreadcrumbs));
        return breadcrumb;
    }

    /// <summary>
    /// Updates the current navigation state without touching the breadcrumb stack.
    /// </summary>
    public void SetCurrent(HistoricalViewLevel level, DateTime date)
    {
        CurrentLevel = level;
        CurrentDate = date;
    }

    /// <summary>
    /// Clears the breadcrumb stack and sets the current view to the supplied values.
    /// </summary>
    public void Reset(HistoricalViewLevel level, DateTime date)
    {
        _breadcrumbStack.Clear();
        OnPropertyChanged(nameof(HasBreadcrumbs));
        SetCurrent(level, date);
    }
}

File: kotlin/shared/src/commonMain/kotlin/com/wellnesswingman/domain/navigation/HistoricalNavigationContext.kt
package com.wellnesswingman.domain.navigation

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.datetime.Clock
import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlinx.datetime.todayIn

/**
 * State holder for navigation context.
 */
data class NavigationState(
    val currentLevel: HistoricalViewLevel = HistoricalViewLevel.TODAY,
    val currentDate: LocalDate = Clock.System.todayIn(TimeZone.currentSystemDefault()),
    val breadcrumbs: List<NavigationBreadcrumb> = emptyList()
) {
    val hasBreadcrumbs: Boolean get() = breadcrumbs.isNotEmpty()
}

/**
 * Tracks the currently displayed historical view and maintains the breadcrumb stack
 * used to restore the previous view when navigating backwards.
 */
class HistoricalNavigationContext {

    private val _state = MutableStateFlow(NavigationState())
    val state: StateFlow<NavigationState> = _state.asStateFlow()

    val currentLevel: HistoricalViewLevel get() = _state.value.currentLevel
    val currentDate: LocalDate get() = _state.value.currentDate
    val hasBreadcrumbs: Boolean get() = _state.value.hasBreadcrumbs

    /**
     * Creates a breadcrumb for the current state so the caller can push it before changing levels.
     */
    fun createCurrentBreadcrumb(label: String? = null): NavigationBreadcrumb {
        return NavigationBreadcrumb(
            level = currentLevel,
            date = currentDate,
            label = label ?: NavigationBreadcrumb(currentLevel, currentDate).label
        )
    }

    /**
     * Returns the most recent breadcrumb without modifying the stack.
     */
    fun peekBreadcrumb(): NavigationBreadcrumb? {
        return _state.value.breadcrumbs.lastOrNull()
    }

    /**
     * Pushes a breadcrumb representing the view the user is drilling down from.
     */
    fun pushBreadcrumb(breadcrumb: NavigationBreadcrumb) {
        _state.update { state ->
            state.copy(breadcrumbs = state.breadcrumbs + breadcrumb)
        }
    }

    /**
     * Pops the most recent breadcrumb when the user navigates back up the hierarchy.
     */
    fun popBreadcrumb(): NavigationBreadcrumb? {
        val breadcrumb = _state.value.breadcrumbs.lastOrNull() ?: return null
        _state.update { state ->
            state.copy(breadcrumbs = state.breadcrumbs.dropLast(1))
        }
        return breadcrumb
    }

    /**
     * Updates the current navigation state without touching the breadcrumb stack.
     */
    fun setCurrent(level: HistoricalViewLevel, date: LocalDate) {
        _state.update { state ->
            state.copy(currentLevel = level, currentDate = date)
        }
    }

    /**
     * Clears the breadcrumb stack and sets the current view to the supplied values.
     */
    fun reset(level: HistoricalViewLevel, date: LocalDate) {
        _state.update {
            NavigationState(
                currentLevel = level,
                currentDate = date,
                breadcrumbs = emptyList()
            )
        }
    }

    /**
     * Resets to today's view.
     */
    fun resetToToday() {
        val today = Clock.System.todayIn(TimeZone.currentSystemDefault())
        reset(HistoricalViewLevel.TODAY, today)
    }
}
